modules/apps/portal-scheduler/portal-scheduler-quartz/src/main/java/com/liferay/portal/scheduler/quartz/internal/QuartzSchedulerEngine.java

```diff
@@ -796,20 +842,64 @@ private void _updateJobState(
 
 	private volatile boolean _schedulerEngineEnabled;
 
-	private class SchedulerListenerImpl extends SchedulerListenerSupport {
+	private static class ConnectionProviderImpl implements ConnectionProvider {
+
+		public ConnectionProviderImpl(long companyId) {
+			_companyId = companyId;
+		}
+
+		@Override
+		public Connection getConnection() {
+			Connection connection = null;
+
+			try (SafeCloseable safeCloseable =
+					CompanyThreadLocal.setWithSafeCloseable(_companyId)) {
```

Comment:
@kevhlee if we use SafeClosable here we won't keep the CompanyThreadLocal set when we create the statements and set the catalog based on the CompanyThreadLocal in com.liferay.portal.db.partition.DBPartitionUtil#_getConnectionWrapper

So we should use just com.liferay.portal.kernel.security.auth.CompanyThreadLocal#setCompanyId

But we have to be sure this java thread is only use for one company scheduler, can you confirm that?

Reply:
Hey @achaparro, I will try to confirm.

Reply:
Hey @achaparro, regarding your question:

>But we have to be sure this java thread is only used for one company scheduler, can you confirm that?

That should not be a problem based on what I found. So, there are multiple threads for each persisted scheduler instance (e.g. one that handles misfires: `MisfireHandler`) and they will never be used for other persisted scheduler instances. So using just `CompanyThreadLocal.setCompanyId()` should be sufficient.

```log
2023-10-04 01:47:36.472 ERROR [PersistedQuartzSchedulerInstance@20001_QuartzSchedulerThread][QuartzSchedulerEngine:838] Current thread: Thread[PersistedQuartzSchedulerInstance@20001_QuartzSchedulerThread,5,QuartzScheduler:PersistedQuartzSchedulerInstance@20001]
2023-10-04 01:47:38.871 ERROR [PersistedQuartzSchedulerInstance@32675_QuartzSchedulerThread][QuartzSchedulerEngine:838] Current thread: Thread[PersistedQuartzSchedulerInstance@32675_QuartzSchedulerThread,5,QuartzScheduler:PersistedQuartzSchedulerInstance@32675]
2023-10-04 01:48:12.540 ERROR [QuartzScheduler_PersistedQuartzSchedulerInstance@20001-NON_CLUSTERED_MisfireHandler][QuartzSchedulerEngine:838] Current thread: Thread[QuartzScheduler_PersistedQuartzSchedulerInstance@20001-NON_CLUSTERED_MisfireHandler,5,main]
2023-10-04 01:48:12.579 ERROR [QuartzScheduler_PersistedQuartzSchedulerInstance@32675-NON_CLUSTERED_MisfireHandler][QuartzSchedulerEngine:838] Current thread: Thread[QuartzScheduler_PersistedQuartzSchedulerInstance@32675-NON_CLUSTERED_MisfireHandler,5,main]
```